{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting tensorflow_transform\n",
      "  Using cached tensorflow_transform-1.11.0-py3-none-any.whl (446 kB)\n",
      "Collecting pyarrow<7,>=6\n",
      "  Using cached pyarrow-6.0.1-cp310-cp310-macosx_10_13_x86_64.whl (19.2 MB)\n",
      "Collecting tensorflow-metadata<1.12.0,>=1.11.0\n",
      "  Using cached tensorflow_metadata-1.11.0-py3-none-any.whl (52 kB)\n",
      "Collecting tensorflow_transform\n",
      "  Using cached tensorflow_transform-1.10.1-py3-none-any.whl (439 kB)\n",
      "Requirement already satisfied: apache-beam[gcp]<3,>=2.40 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow_transform) (2.43.0)\n",
      "Collecting tensorflow-metadata<1.11.0,>=1.10.0\n",
      "  Using cached tensorflow_metadata-1.10.0-py3-none-any.whl (50 kB)\n",
      "Requirement already satisfied: absl-py<2.0.0,>=0.9 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow_transform) (1.3.0)\n",
      "Requirement already satisfied: protobuf<4,>=3.13 in /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages (from tensorflow_transform) (3.19.4)\n",
      "Collecting tensorflow_transform\n",
      "  Using cached tensorflow_transform-1.10.0-py3-none-any.whl (439 kB)\n",
      "  Using cached tensorflow_transform-1.9.0-py3-none-any.whl (436 kB)\n",
      "  Using cached tensorflow_transform-1.8.0-py3-none-any.whl (435 kB)\n",
      "  Using cached tensorflow_transform-1.7.0-py3-none-any.whl (433 kB)\n",
      "Collecting pyarrow<6,>=1\n",
      "  Using cached pyarrow-5.0.0.tar.gz (739 kB)\n",
      "  Installing build dependencies ... \u001b[?25lerror\n",
      "  \u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "  \n",
      "  \u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "  \u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "  \u001b[31m╰─>\u001b[0m \u001b[31m[214 lines of output]\u001b[0m\n",
      "  \u001b[31m   \u001b[0m Ignoring numpy: markers 'python_version < \"3.9\"' don't match your environment\n",
      "  \u001b[31m   \u001b[0m Collecting cython>=0.29\n",
      "  \u001b[31m   \u001b[0m   Using cached Cython-0.29.32-py2.py3-none-any.whl (986 kB)\n",
      "  \u001b[31m   \u001b[0m Collecting numpy==1.19.4\n",
      "  \u001b[31m   \u001b[0m   Using cached numpy-1.19.4.zip (7.3 MB)\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: started\n",
      "  \u001b[31m   \u001b[0m   Installing build dependencies: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: started\n",
      "  \u001b[31m   \u001b[0m   Getting requirements to build wheel: finished with status 'done'\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): started\n",
      "  \u001b[31m   \u001b[0m   Preparing metadata (pyproject.toml): finished with status 'error'\n",
      "  \u001b[31m   \u001b[0m   error: subprocess-exited-with-error\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   × Preparing metadata (pyproject.toml) did not run successfully.\n",
      "  \u001b[31m   \u001b[0m   │ exit code: 1\n",
      "  \u001b[31m   \u001b[0m   ╰─> [188 lines of output]\n",
      "  \u001b[31m   \u001b[0m       setup.py:67: RuntimeWarning: NumPy 1.19.4 may not yet support Python 3.10.\n",
      "  \u001b[31m   \u001b[0m         warnings.warn(\n",
      "  \u001b[31m   \u001b[0m       Running from numpy source directory.\n",
      "  \u001b[31m   \u001b[0m       setup.py:480: UserWarning: Unrecognized setuptools command, proceeding with generating Cython sources and expanding templates\n",
      "  \u001b[31m   \u001b[0m         run_build = parse_setuppy_commands()\n",
      "  \u001b[31m   \u001b[0m       Cythonizing sources\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_bounded_integers.pxd.in\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_philox.pyx\n",
      "  \u001b[31m   \u001b[0m       /private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/tools/cythonize.py:67: DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives\n",
      "  \u001b[31m   \u001b[0m         from distutils.version import LooseVersion\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_bounded_integers.pyx.in\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_sfc64.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_mt19937.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/bit_generator.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/mtrand.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_generator.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_pcg64.pyx\n",
      "  \u001b[31m   \u001b[0m       Processing numpy/random/_common.pyx\n",
      "  \u001b[31m   \u001b[0m       blas_opt_info:\n",
      "  \u001b[31m   \u001b[0m       blas_mkl_info:\n",
      "  \u001b[31m   \u001b[0m       customize UnixCCompiler\n",
      "  \u001b[31m   \u001b[0m         libraries mkl_rt not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       blis_info:\n",
      "  \u001b[31m   \u001b[0m         libraries blis not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       openblas_info:\n",
      "  \u001b[31m   \u001b[0m         libraries openblas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_3_10_blas_threads_info:\n",
      "  \u001b[31m   \u001b[0m       Setting PTATLAS=ATLAS\n",
      "  \u001b[31m   \u001b[0m         libraries tatlas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_3_10_blas_info:\n",
      "  \u001b[31m   \u001b[0m         libraries satlas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_blas_threads_info:\n",
      "  \u001b[31m   \u001b[0m       Setting PTATLAS=ATLAS\n",
      "  \u001b[31m   \u001b[0m         libraries ptf77blas,ptcblas,atlas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_blas_info:\n",
      "  \u001b[31m   \u001b[0m         libraries f77blas,cblas,atlas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       accelerate_info:\n",
      "  \u001b[31m   \u001b[0m         libraries accelerate not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m       Library accelerate was not found. Ignoring\n",
      "  \u001b[31m   \u001b[0m         libraries veclib not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m       Library veclib was not found. Ignoring\n",
      "  \u001b[31m   \u001b[0m         FOUND:\n",
      "  \u001b[31m   \u001b[0m           extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "  \u001b[31m   \u001b[0m           extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "  \u001b[31m   \u001b[0m           define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         FOUND:\n",
      "  \u001b[31m   \u001b[0m           extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "  \u001b[31m   \u001b[0m           extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "  \u001b[31m   \u001b[0m           define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       non-existing path in 'numpy/distutils': 'site.cfg'\n",
      "  \u001b[31m   \u001b[0m       lapack_opt_info:\n",
      "  \u001b[31m   \u001b[0m       lapack_mkl_info:\n",
      "  \u001b[31m   \u001b[0m         libraries mkl_rt not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       openblas_lapack_info:\n",
      "  \u001b[31m   \u001b[0m         libraries openblas not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       openblas_clapack_info:\n",
      "  \u001b[31m   \u001b[0m         libraries openblas,lapack not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       flame_info:\n",
      "  \u001b[31m   \u001b[0m         libraries flame not found in ['/Library/Frameworks/Python.framework/Versions/3.10/lib', '/usr/local/lib', '/usr/lib']\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_3_10_threads_info:\n",
      "  \u001b[31m   \u001b[0m       Setting PTATLAS=ATLAS\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries tatlas,tatlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries tatlas,tatlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m         libraries tatlas,tatlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m       <class 'numpy.distutils.system_info.atlas_3_10_threads_info'>\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_3_10_info:\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries satlas,satlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries satlas,satlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m         libraries satlas,satlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m       <class 'numpy.distutils.system_info.atlas_3_10_info'>\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_threads_info:\n",
      "  \u001b[31m   \u001b[0m       Setting PTATLAS=ATLAS\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries ptf77blas,ptcblas,atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries ptf77blas,ptcblas,atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m         libraries ptf77blas,ptcblas,atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m       <class 'numpy.distutils.system_info.atlas_threads_info'>\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       atlas_info:\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries f77blas,cblas,atlas not found in /Library/Frameworks/Python.framework/Versions/3.10/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries f77blas,cblas,atlas not found in /usr/local/lib\n",
      "  \u001b[31m   \u001b[0m         libraries lapack_atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m         libraries f77blas,cblas,atlas not found in /usr/lib\n",
      "  \u001b[31m   \u001b[0m       <class 'numpy.distutils.system_info.atlas_info'>\n",
      "  \u001b[31m   \u001b[0m         NOT AVAILABLE\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m         FOUND:\n",
      "  \u001b[31m   \u001b[0m           extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']\n",
      "  \u001b[31m   \u001b[0m           extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']\n",
      "  \u001b[31m   \u001b[0m           define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m       /Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/dist.py:274: UserWarning: Unknown distribution option: 'define_macros'\n",
      "  \u001b[31m   \u001b[0m         warnings.warn(msg)\n",
      "  \u001b[31m   \u001b[0m       running dist_info\n",
      "  \u001b[31m   \u001b[0m       running build_src\n",
      "  \u001b[31m   \u001b[0m       build_src\n",
      "  \u001b[31m   \u001b[0m       building py_modules sources\n",
      "  \u001b[31m   \u001b[0m       creating build\n",
      "  \u001b[31m   \u001b[0m       creating build/src.macosx-10.9-universal2-3.10\n",
      "  \u001b[31m   \u001b[0m       creating build/src.macosx-10.9-universal2-3.10/numpy\n",
      "  \u001b[31m   \u001b[0m       creating build/src.macosx-10.9-universal2-3.10/numpy/distutils\n",
      "  \u001b[31m   \u001b[0m       building library \"npymath\" sources\n",
      "  \u001b[31m   \u001b[0m       xcrun: error: invalid active developer path (/Library/Developer/CommandLineTools), missing xcrun at: /Library/Developer/CommandLineTools/usr/bin/xcrun\n",
      "  \u001b[31m   \u001b[0m       Traceback (most recent call last):\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py\", line 351, in <module>\n",
      "  \u001b[31m   \u001b[0m           main()\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py\", line 333, in main\n",
      "  \u001b[31m   \u001b[0m           json_out['return_val'] = hook(**hook_input['kwargs'])\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py\", line 152, in prepare_metadata_for_build_wheel\n",
      "  \u001b[31m   \u001b[0m           return hook(metadata_directory, config_settings)\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-build-env-991lyu4x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py\", line 157, in prepare_metadata_for_build_wheel\n",
      "  \u001b[31m   \u001b[0m           self.run_setup()\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-build-env-991lyu4x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py\", line 248, in run_setup\n",
      "  \u001b[31m   \u001b[0m           super(_BuildMetaLegacyBackend,\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-build-env-991lyu4x/overlay/lib/python3.10/site-packages/setuptools/build_meta.py\", line 142, in run_setup\n",
      "  \u001b[31m   \u001b[0m           exec(compile(code, __file__, 'exec'), locals())\n",
      "  \u001b[31m   \u001b[0m         File \"setup.py\", line 508, in <module>\n",
      "  \u001b[31m   \u001b[0m           setup_package()\n",
      "  \u001b[31m   \u001b[0m         File \"setup.py\", line 500, in setup_package\n",
      "  \u001b[31m   \u001b[0m           setup(**metadata)\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/core.py\", line 169, in setup\n",
      "  \u001b[31m   \u001b[0m           return old_setup(**new_attr)\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-build-env-991lyu4x/overlay/lib/python3.10/site-packages/setuptools/__init__.py\", line 165, in setup\n",
      "  \u001b[31m   \u001b[0m           return distutils.core.setup(**attrs)\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/core.py\", line 148, in setup\n",
      "  \u001b[31m   \u001b[0m           dist.run_commands()\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/dist.py\", line 966, in run_commands\n",
      "  \u001b[31m   \u001b[0m           self.run_command(cmd)\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/dist.py\", line 985, in run_command\n",
      "  \u001b[31m   \u001b[0m           cmd_obj.run()\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-build-env-991lyu4x/overlay/lib/python3.10/site-packages/setuptools/command/dist_info.py\", line 31, in run\n",
      "  \u001b[31m   \u001b[0m           egg_info.run()\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/command/egg_info.py\", line 24, in run\n",
      "  \u001b[31m   \u001b[0m           self.run_command(\"build_src\")\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/cmd.py\", line 313, in run_command\n",
      "  \u001b[31m   \u001b[0m           self.distribution.run_command(command)\n",
      "  \u001b[31m   \u001b[0m         File \"/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/distutils/dist.py\", line 985, in run_command\n",
      "  \u001b[31m   \u001b[0m           cmd_obj.run()\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/command/build_src.py\", line 144, in run\n",
      "  \u001b[31m   \u001b[0m           self.build_sources()\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/command/build_src.py\", line 155, in build_sources\n",
      "  \u001b[31m   \u001b[0m           self.build_library_sources(*libname_info)\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/command/build_src.py\", line 288, in build_library_sources\n",
      "  \u001b[31m   \u001b[0m           sources = self.generate_sources(sources, (lib_name, build_info))\n",
      "  \u001b[31m   \u001b[0m         File \"/private/var/folders/pw/96pjsf2s6q73cvqqql90pw600000gn/T/pip-install-gknblv0s/numpy_66d8efc4e69b4e9599c03b3aa17375d5/numpy/distutils/command/build_src.py\", line 378, in generate_sources\n",
      "  \u001b[31m   \u001b[0m           source = func(extension, build_dir)\n",
      "  \u001b[31m   \u001b[0m         File \"numpy/core/setup.py\", line 663, in get_mathlib_info\n",
      "  \u001b[31m   \u001b[0m           raise RuntimeError(\"Broken toolchain: cannot link a simple C program\")\n",
      "  \u001b[31m   \u001b[0m       RuntimeError: Broken toolchain: cannot link a simple C program\n",
      "  \u001b[31m   \u001b[0m       [end of output]\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m   note: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "  \u001b[31m   \u001b[0m error: metadata-generation-failed\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m × Encountered error while generating package metadata.\n",
      "  \u001b[31m   \u001b[0m ╰─> See above for output.\n",
      "  \u001b[31m   \u001b[0m \n",
      "  \u001b[31m   \u001b[0m note: This is an issue with the package mentioned above, not pip.\n",
      "  \u001b[31m   \u001b[0m hint: See above for details.\n",
      "  \u001b[31m   \u001b[0m \u001b[31m[end of output]\u001b[0m\n",
      "  \n",
      "  \u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n",
      "\u001b[?25h\u001b[1;31merror\u001b[0m: \u001b[1msubprocess-exited-with-error\u001b[0m\n",
      "\n",
      "\u001b[31m×\u001b[0m \u001b[32mpip subprocess to install build dependencies\u001b[0m did not run successfully.\n",
      "\u001b[31m│\u001b[0m exit code: \u001b[1;36m1\u001b[0m\n",
      "\u001b[31m╰─>\u001b[0m See above for output.\n",
      "\n",
      "\u001b[1;35mnote\u001b[0m: This error originates from a subprocess, and is likely not a problem with pip.\n"
     ]
    }
   ],
   "source": [
    "!pip install -U tensorflow_transform"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Overview\n",
    "\n",
    "Author: stephankokkas\n",
    "\n",
    "This notebook defines a pipeline that tasks an input directory of audio files and converts them to images using mel-spectrogram transofrmation and preprocessing techniques."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'tensorflow_transform'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Input \u001b[0;32mIn [28]\u001b[0m, in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m# tensorflow support\u001b[39;00m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtensorflow\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mtf\u001b[39;00m\n\u001b[0;32m---> 21\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtensorflow_transform\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mtft\u001b[39;00m\n\u001b[1;32m     22\u001b[0m \u001b[38;5;66;03m#import tensorflow_io as tfio\u001b[39;00m\n\u001b[1;32m     23\u001b[0m \u001b[38;5;66;03m#from tensorflow.contrib.framework.python.ops import audio_ops\u001b[39;00m\n\u001b[1;32m     24\u001b[0m \n\u001b[1;32m     25\u001b[0m \u001b[38;5;66;03m# scipy\u001b[39;00m\n\u001b[1;32m     26\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mscipy\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'tensorflow_transform'"
     ]
    }
   ],
   "source": [
    "# disable warnings to tidy up output\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "# some basic libraries \n",
    "from platform import python_version\n",
    "#import pandas as pd\n",
    "#import seaborn as sns\n",
    "import numpy as np\n",
    "import os\n",
    "import random\n",
    "from os import listdir\n",
    "from os.path import isfile, join\n",
    "\n",
    "\n",
    "# plot support\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "# tensorflow support\n",
    "import tensorflow as tf\n",
    "import tensorflow_transform as tft\n",
    "#import tensorflow_io as tfio\n",
    "#from tensorflow.contrib.framework.python.ops import audio_ops\n",
    "\n",
    "# scipy\n",
    "import scipy\n",
    "\n",
    "# turn off tensorflow warnings\n",
    "tf.get_logger().setLevel('ERROR')\n",
    "\n",
    "# turn off absl warnings\n",
    "import absl.logging\n",
    "absl.logging.set_verbosity(absl.logging.ERROR)\n",
    "\n",
    "# print system information\n",
    "print('Python Version     : ', python_version())\n",
    "print('TensorFlow Version : ', tf.__version__)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "zsh:1: command not found: python\n"
     ]
    }
   ],
   "source": [
    "!python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# below code adapted from:\n",
    "# https://keras.io/getting_started/faq/#how-can-i-obtain-reproducible-results-using-keras-during-development\n",
    "\n",
    "# The below is necessary for starting Numpy generated random numbers\n",
    "# in a well-defined initial state.\n",
    "np.random.seed(123)\n",
    "\n",
    "# The below is necessary for starting core Python generated random numbers\n",
    "# in a well-defined state.\n",
    "random.seed(123)\n",
    "\n",
    "# The below set_seed() will make random number generation\n",
    "# in the TensorFlow backend have a well-defined initial state.\n",
    "# For further details, see:\n",
    "# https://www.tensorflow.org/api_docs/python/tf/random/set_seed\n",
    "tf.random.set_seed(1234)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set system parameters\n",
    "DATASET_PATH  = 'C:/Users/Andrew/OneDrive - Deakin University/DataSets/birdclef2022/'\n",
    "FILE_FORMAT = ''\n",
    "\n",
    "SAMPLE_RATE   = 32000   # all the samples are converted to bit rate of 32000 (Samples/Second)\n",
    "MIN_FREQUENCY = 16      # minimum frequency (Hz) for the Fast Fourier Transform related functions\n",
    "MAX_FREQUENCY = 4096*3  # minimum frequency (Hz) for the Fast Fourier Transform related functions\n",
    "HOP_LENGTH    = 128     # the number of samples to slide spectrogram window along the audio samples\n",
    "NUMBER_FFT    = 2048    # the number of FFT to execute within a single spectrogram window\n",
    "NUMBER_MELS   = 128     # the number of Mel-Spectrogram groups to split the frequency dimension\n",
    "CLIP_LENGTH   = 5      # only look at 10 seconds of clip at the start of loaded audio file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "aee8b7b246df8f9039afb4144a1f6fd8d2ca17a180786b69acc140d282b71a49"
  },
  "kernelspec": {
   "display_name": "Python 3.10.0 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
